module/mempool/herocache/backdata/heropool/pool_test.go

```diff
@@ -215,6 +231,97 @@ func TestInvalidateEntity(t *testing.T) {
 	}
 }
 
+// TestAddAndRemoveEntities checks health of heroPool for scenario where entitites are stored and removed in a predetermined order.
+// LRUEjection, NoEjection and RandomEjection are tested. RandomEjection doesn't allow to provide a final state of the pool to check.
+func TestAddAndRemoveEntities(t *testing.T) {
+	for _, tc := range []struct {
+		limit             uint32              // capacity of the pool
+		entityCount       uint32              // total entities to be stored
+		ejectionMode      EjectionMode        // ejection mode
+		operationsIndexes []OperationAndIndex // operation to perform on an entity stored under the given index,
+		// where index ranges from 0 to entityCount
+		finalEntitiesInThePool []uint // indexes of entities residing in the pool at the end of the test
+	}{
+		{
+			limit:                  2,
+			entityCount:            5,
+			ejectionMode:           LRUEjection,
+			operationsIndexes:      []OperationAndIndex{{Add, 0}, {Add, 0}, {Add, 1}, {Add, 2}, {Add, 3}, {Remove, 2}, {Add, 2}, {Remove, 2}},
```

Comment:
Please consider generating these operations at scale using a test case generator function.

Reply:
The array has to be hardcoded in the test or read from a separate asset file. I could write a script that would use some random generator and write an array of say 1000 operations. Then, I think it would be good to save it as an asset in a txt format in the same package.
(I would not suggest to generate operations each time as test is executed as then test will loose consistency. I mean if we go to 10000 operations may be we could say that random generation is ok... I would like to know your opinion on that matter.)

Reply:
I would advocate for employing a randomized one-time generation approach, as it has the potential to unveil and capture hidden edge cases and unanticipated bugs. While using a script to generate the test cases may seem like a viable option, it often becomes cumbersome to maintain. Furthermore, it can be challenging to ensure that the generation of fresh randomized test cases is consistently incorporated into every CI task. Utilizing randomized one-time generation could provide a more robust and automated solution, enhancing the quality and reliability of the tests.

Reply:
I think a randomized test may overlook certain edge cases too. For example in current randomized test pool is never close to empty. The probability of "Add" and "Remove" are equal. But there are 1000 entities.
In the beginning we get something like Add 0, Remove 900, Add 500, Remove 3

So, although probabilities of Add and Remove are equal , Removes do nothing for quiet a while ... Remove start to be active when pool has considerable amount of entities.

I am not saying we don't need test with many random operation but I truly believe such a test is complementary to a human crafted logical test and can't be a replacement.

Reply:
You can incorporate both randomness and bias into the test enhances its robustness. For instance, you can randomly choose between `Add` and `Remove` operations with respective probabilities of `0.8` and `0.2`, diverging from an equal chance. When `Add` is selected, the test should generate a _random_ and _unique_ entity, add it to the `Pool`, and maintain a record of it for subsequent validation. Conversely, if `Remove` is chosen, the test should randomly select an entity that was previously added and remove it from the `Pool`. Below is the refined pseudocode in Golang:

```go
	rand.Seed(time.Now().UnixNano())

	pool := []Entity{}
	addedEntities := []Entity{}

	const numIterations = 1000
	for i := 0; i < numIterations; i++ {
		// Choose between Add and Remove with a probability of 0.8 and 0.2 respectively
		if rand.Float32() < 0.8 || len(addedEntities) == 0 {
			// TODO: Create a random entity and add it to the pool
		} else {
			// Randomly select an index of an entity to remove
			indexToRemove := rand.Intn(len(addedEntities))

			// Remove the selected entity from the pool
			entityToRemove := addedEntities[indexToRemove]
			// TODO: Remove logic from pool
		}
	}
```

Reply:
Sure, many things are possible. I'll implement your suggestion then.

Reply:
PS: 0.8 and 0.2 would keep the pool almost full as in the current test implementation with 0.5 probabilities.

Reply:
I have implemented your suggestion with a difference that I've used a hash map instead of an array for added entities. 

I map added flow.id to an index of a flow entity in an array of entities created by ```unittest.EntityListFixture```. You'll see in the code several usages of this map ```addedEntities``` but as test now is pretty fast I could also change it to an array. I think it wont be a problem. 

As for probabilities, I made them configurable. I believe 0.8 for add and 0.2 for remove would lead to the similar test issue that I have described previously. Test will test almost always a full pool. So, it's not to say that this test scenario is bad but just to highlight that hand crafted test would allow to test some precisely defined edge test cases. Also, as we cannot construct an 'end' state of the pool by hand we are relying on the fact that checks are coded properly. That's another disadvantage compared to a classical unit test. What if there are 2 errors - one in code and one in a data structure that maintains data about hat test think there should be in the pool. Anyway, I concur that test with many random operation is a good idea. So, I am happy to add this test in the current form too.

As for a random seed, it indeed allows to cover more cases with time, but if some corner case is discovered only with a particular seed one can easily integrate a change with an error and then only after some time get an error from CI. 
Nevertheless, as we see current test as a complementary test to existing ones and there are many operations, it is probably a minor thing.